### DISCUSSION TITLE
Replacing primary firmware image with secondary image on basis of latest version

### DISCUSSION THREAD

**Discussion by waqarsadiq** [mcu-tools/mcuboot#1267] [General]
Hi,
I am creating two images both with using the option imgtool.py sign --overwrite-only ........
Primary image goes to internal flash of the device (PRI-IMG) has version 1.0.0
Secondary goes to the external flash of the device (SEC-IMG) has version 2.0.0
Goal : Device boots and checks if a SEC-IMG is present in external flash with greater version number and if it is there and is also latest one then it should overwrite the PRI-IMG in internal flash with SEC-iMG 
When MCUboot boots it scans and picks up the two images but doesn't perform an update.
I can force an update using boot_set_pending_multi(). Are there any configurations in MCUboot and imgtool through which I can make MCUboot to do this automatically for me without writing any additional logic in my bootloader application?

**Comment by Olstyle** [comment]
As far as I know this is already the behavior if you build an image with a valid header and trailer. But the later only happens if you use `--pad` in imgtool.

**Comment by d3zd3z** [comment]
> As far as I know this is already the behavior if you build an image with a valid header and trailer. But the later only happens if you use --pad in imgtool.

imgtool only does this with `--pad` because the indicator for the upgrade is at the end of the image, and without padding, imgtool isn't aware even of the size of the image.

Currently, this is done by writing the magic value at the end of the slot, but using either img tool or the api is recommended since then the mechanism will continue to work in light of future changes to the trailer area.

**Comment by waqarsadiq** [comment]
Right. @d3zd3z 
If I use the --pad option with img-tool then how can I modify some certain entries of trailer e.g. "Swap info" or "Image ok"?
I couldn't find any documentation for options which can be passed with --pad.

**Reply by Olstyle** [reply]
Those states you are referring to are working entries used by mcuboot internally. An image made with imgtool will always be ok (why would you want anything else?) and not swapped. But you can change mandate it is already confirmed with `--confirmed` and the swap type to overwrite with `--overwrite-only`. From the python code it looks like `--confirmed` will actually lead to padding being enabled as well because that info needs the trailer.

**Reply by d3zd3z** [reply]
This is correct, only the magic value and "image ok" are meaningful fields outside of the internals of MCUboot.  `--pad` will set the magic value, and `--confirmed` will set the image ok field, so that instead of a swap, the image is simply copied into place (what actually happens depends on configuration).

**Comment by chanakya-macharla** [comment]
Hello Techies, 

We have used the same options i.e. --pad  and --overwrite-only for sign the binary and keeping in secondary slot for upgrading image and rebooted it, we have observed the image is upgraded but we got the info messages as 

**"[INF] Image index: 0, Swap type: test
[INF] Image 0 upgrade secondary slot -> primary slot"**

below command we have used to sign the binary

**%python% %imgtool% sign -k %key% -e little -S %image_size% -M %image_sectors% -H 0x400 --pad-header %option% -v %version% -s %security_counter% --overwrite-only --pad --align %val% --max-align %val% firmware.bin firmwareinit.bin**


our issue is why **"Swap type: test"** for overwrite mode. do we need to add or remove any options?


Thanks & regards 
Chanakya M
